test(CL Fees): test values for `GetFeeGrowthOutside`

mattverse opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: https://github.com/osmosis-labs/osmosis/issues/3958

## What is the purpose of the change

Adds and uncomments testing vectors for `GetFeeGrowthOutside`


## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -74,41 +74,44 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+		"fee growth outside with tick initialized": {
+			poolSetup:                true,
+			tickSetup:                true,
+			expectedFeeGrowthOutside: defaultAccumCoins,
```
We need to have more tests for various cases of what fee growth outside is.

For example, assume we have a position at ticks 10 and 20. As a swap is completed, we cross the ticks from left to right. Both ticks are crossed and the current tick stops at tick 25. As a result, both tick 10 and tick 20 are updated with the `feeGrowthGlobal`.

We need to simulate and test all the edge cases such as:
- crossing left to right
   *  the current price stays before the lower tick so none of the tick accumulators are updated
   *  the current price stays in-between the lower and upper tick so only lower tick is updated
   *  the current price stays after the upper tick so both lower and upper tick accumulators are updated
- crossing right to left
   * similar test cases to left to right but in the other direction.


I did a similar test for `collectFees` in https://github.com/osmosis-labs/osmosis/pull/3893
Test link:
https://github.com/osmosis-labs/osmosis/blob/e1e8a4624fdd844fdd8536de13ca8924a3c51de2/x/concentrated-liquidity/fees_test.go#L400

We need something similar but more focused on `feeGrowthOutside` specifically. I think it's important that we test all the edge cases at a lower level of abstraction. Otherwise, if we miss something, it will be close to impossible to debug at a higher level of abstraction (e.g. in swaps) so we will have to go back here.

mattverse commented:
@p0mvn  addressed comments, referenced the specific cases you linked to this case, but specifically focusing on testing the GetFeeGrowthOutside testings

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
-		// 	feeAccumulatorSetup: true,
-		// 	expectedError:       false,
-		// },
+		"single swap left -> right: 2 ticks, one share - current price > upper tick": {
```
Could you please add an explanation comment similar to what you have for left-to-right tests cases?

mattverse replied:
done!

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+			expectedFeeGrowthOutside:  defaultAccumCoins,
+			expectedError:             false,
+		},
+		"single swap left -> right: 3 ticks, two shares - current price > upper tick": {
```
```suggestion
		"single swap left -> right: 3 ticks, two shares - current tick > upper tick": {
```

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
-		// 	feeAccumulatorSetup: true,
-		// 	expectedError:       false,
-		// },
+		"single swap left -> right: 2 ticks, one share - current price > upper tick": {
```
```suggestion
		"single swap left -> right: 2 ticks, one share - current tick > upper tick": {
```

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+			expectedFeeGrowthOutside:  defaultAccumCoins,
+			expectedError:             false,
+		},
+		"single swap left -> right: 2 ticks, one share - current price == upper tick": {
```
```suggestion
		"single swap left -> right: 2 ticks, one share - current tick == lower tick": {
```

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+			globalFeeGrowth:           sdk.NewDecCoin(ETH, sdk.NewInt(10)),
+			expectedFeeGrowthOutside:  defaultAccumCoins,
+			expectedError:             false,
```
Suggest also having the following for left to right test cases:
- `current tick == upper tick`
- `current tick < lower tick`

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+		// In this case, all fees must have been accrued inside the tick
+		// Since we track fees accrued below a tick, both upper and lower position
+		// ticks are zero
+		"single swap right -> left: 2 ticks, one share - current price == lower tick": {
```
```suggestion
		"single swap right -> left: 2 ticks, one share - current tick == lower tick": {
```

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+			expectedFeeGrowthOutside:  defaultAccumCoins,
+			expectedError:             false,
+		},
+		"single swap right -> left: 2 ticks, one share - current price < lower tick": {
```
```suggestion
		"single swap right -> left: 2 ticks, one share - current tick < lower tick": {
```

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,111 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+			globalFeeGrowth:           sdk.NewDecCoin(ETH, sdk.NewInt(10)),
+			expectedFeeGrowthOutside:  defaultAccumCoins,
+			expectedError:             false,
+		},
```
suggest also having the following for right to left test cases:
- `current tick == upper tick`
- `current tick > upper tick`

p0mvn replied:
I think that `current tick > upper tick` is still missing

mattverse replied:
added "single swap right -> left: 2 ticks, one share - current tick > upper tick":

p0mvn reviewed (commented):
Looks great. Requesting more test cases, lgtm otherwise

mattverse commented:
@p0mvn  I actually have a proposal regarding these test cases, the more I think about these cases and the test cases you mentioned(https://github.com/osmosis-labs/osmosis/pull/3979#discussion_r1067678790) as well as for the test cases`collectFees`, I feel like these test cases are closer to integration tests for specific edge cases and scenarios, more than unit testing. I would like to propose reverting these test cases back to `https://github.com/osmosis-labs/osmosis/pull/3979/commits/d6a2b4f1910bb42ed680ed574a635125513d678c` and then creating new test cases that mentions this and tests both `collectFees` and `GetFeeGrowthOutside`. What do you think?

p0mvn commented:
> @p0mvn I actually have a proposal regarding these test cases, the more I think about these cases and the test cases you mentioned([#3979 (comment)](https://github.com/osmosis-labs/osmosis/pull/3979#discussion_r1067678790)) as well as for the test cases`collectFees`, I feel like these test cases are closer to integration tests for specific edge cases and scenarios, more than unit testing. I would like to propose reverting these test cases back to [d6a2b4f](https://github.com/osmosis-labs/osmosis/commit/d6a2b4f1910bb42ed680ed574a635125513d678c) and then creating new test cases that includes this and tests both `collectFees` and `GetFeeGrowthOutside` with the scenarios you provided. What do you think?

I would still consider these to be unit tests. The unit test names are framed in terms of swaps because it is easier for us humans to reason about them. That's why it might feel like these are more functional in nature.

I don't really understand what exactly you're proposing with reverting and re-recreating test cases. I still think that we should have the requested test cases to validate that the internal calculations are done correctly at each level of abstraction (`collectFees`, `getFeeGrowthOutside` etc). As long we are tracking/planning to add them in the future, any approach is fine with me.

mattverse commented:
> > @p0mvn I actually have a proposal regarding these test cases, the more I think about these cases and the test cases you mentioned([#3979 (comment)](https://github.com/osmosis-labs/osmosis/pull/3979#discussion_r1067678790)) as well as for the test cases`collectFees`, I feel like these test cases are closer to integration tests for specific edge cases and scenarios, more than unit testing. I would like to propose reverting these test cases back to [d6a2b4f](https://github.com/osmosis-labs/osmosis/commit/d6a2b4f1910bb42ed680ed574a635125513d678c) and then creating new test cases that includes this and tests both `collectFees` and `GetFeeGrowthOutside` with the scenarios you provided. What do you think?
> 
> I would still consider these to be unit tests. The unit test names are framed in terms of swaps because it is easier for us humans to reason about them. That's why it might feel like these are more functional in nature.
> 
> I don't really understand what exactly you're proposing with reverting and re-recreating test cases. I still think that we should have the requested test cases to validate that the internal calculations are done correctly at each level of abstraction (`collectFees`, `getFeeGrowthOutside` etc). As long we are tracking/planning to add them in the future, any approach is fine with me.

Hey @p0mvn! I don't have strong personal opinions regarding this case as well, I'm fine with either both ways. I think it's a  difference in point of view in which way is intuitive.

Decision made to stay as is, addressed your comments as well!

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,159 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+			poolSetup:                 true,
+			lowerTick:                 0,
+			upperTick:                 1,
+			currentTick:               0,
```
It seems to me that here currentTick == lowerTick, not current tick == upper tick like the test case name says

mattverse replied:
good catch, fixed

p0mvn commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,159 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+		// been updated when crossed.
+		// Since we track fees accrued below a tick, upper tick is updated
+		// while lower tick is zero.
+		"single swap left -> right: 2 ticks, one share - current price > upper tick": {
```
```suggestion
		"single swap left -> right: 2 ticks, one share - current tick > upper tick": {
```

AlpinYukseloglu commented on x/concentrated-liquidity/fees_test.go:
```diff
@@ -144,67 +144,159 @@ func (s *KeeperTestSuite) TestInitializeFeeAccumulatorPosition() {
+				s.initializeTick(s.Ctx, tc.upperTick, defaultInitialLiquidity, tc.upperTickFeeGrowthOutside, true)
+				pool.SetCurrentTick(sdk.NewInt(tc.currentTick))
+				s.App.ConcentratedLiquidityKeeper.SetPool(s.Ctx, pool)
+				// poolId = pool.GetId()
```
```suggestion
```

AlpinYukseloglu reviewed (approved):
Would love to fuzz test fees across tick ranges and fee growths, but that might be better suited for a separate PR. This one LGTM!
